### DISCUSSION TITLE
sparse table output

### DISCUSSION THREAD

**Discussion by kiminh** [Qihoo360/tensornet#66] [General] [answered]
i have run the wide and deep model on industrial dataset and got the sparse table output. @thundercatli 

```
sparse_opt = tn.core.AdaGrad(learning_rate=0.01, initial_g2sum=0.1, initial_scale=0.1)

opt_name:AdaGrad
dim:8
5550086314013414980     8       0.0506565       0.0165137       -0.000290567    -0.0096354      -0.00974899     0.0368803       -0.017581  -0.0189105       0.1     1.96    0       0
3064317571301576340     8       -0.0575744      -0.0326009      0.0180476       -0.0123906      0.0242864       -0.0342714      0.0521148  0.0441031        0.1     0.98    0       0
1756588472783864528     8       0.0126326       0.0403345       -0.0335226      0.00557136      0.0392844       -0.0347473      -0.0147189 0.00544515       0.1     3.92    0       0
598407693234333660      8       0.0161784       -0.0303006      -0.0684115      0.02354 0.00463372      -0.0147034      0.0116275       -0.0255773  0.1     0.98    0       0
2807026422257546872     8       0.0213118       -0.000866071    0.012316        0.0103005       -0.0127499      0.00723172      -0.0150036 -0.0837113       0.1     1.96    0       0
6118954042852362792     8       0.00998451      -0.0084689      0.0401853       0.00317581      -0.0709747      -0.0203387      -0.000485013 -0.00477354     0.1     1.96    0       0
4096748474065003872     8       0.00899766      -0.0275911      -0.0161629      -0.00443562     -0.0283328      -0.0747014      -0.00938712 -0.0270782       0.1     16.66   0       9.8
4174304074430002404     8       0.0113422       0.0576163       -0.0221065      -0.0159259      -0.00519418     0.0400519       -0.00232559 0.0979451        0.1     1.96    0       0
```
I'm curious about the meaning of every column, here use_cvm_ is false.

the above has one more column which meaning is not clear, as it's conflict with the example, and i'm not able to figure out its meaning

https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/core/ps/table/sparse_table.cc#L93
https://github.com/Qihoo360/tensornet/blob/master/core/ps/optimizer/ada_grad_kernel.cc#L131
```
void SparseAdaGradValue::SerializeTxt_(std::ostream& os, int dim) {
    os << dim << "\t";
    for (int i = 0; i < dim; i++) {
        os << Weight()[i] << "\t";
    }

    os << g2sum_ << "\t";
    os << show_ << "\t";
    os << no_show_days_ << "\t";
    if (use_cvm_) {
        os << click_;
    }
}
```
https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/tensornet_tools/python/utils.py#L30

**Comment by thundercatli** [comment, marked as answer]
https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/examples/common/feature_column.py#L49

If you use target_columns option, use_cvm_ will be set to true. There will be one more column in sparse output to calculate label count. sparse_ada_grad_schema in tensornet-tools is not update-to-date. Or you don't need to enable target_columns.

**Reply by kiminh** [reply]
> https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/examples/common/feature_column.py#L49
> 
> If you use target_columns option, use_cvm_ will be set to true. There will be one more column in sparse output to calculate label count. sparse_ada_grad_schema in tensornet-tools is not update-to-date. Or you don't need to enable target_columns.

thanks for your reply. 
then for single task prediction task, the sparse output will has one more label column. for multi task (ctr prediction and cvr prediction), not support yet.

https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/tensornet/layers/embedding_features.py#L219
```
        if target_columns and len(target_columns) > 1:
            raise ValueError("For now cvm plugin only support one column, Given: {}".format(target_columns))
```

**Reply by kiminh** [reply]
> https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/examples/common/feature_column.py#L49
> 
> If you use target_columns option, use_cvm_ will be set to true. There will be one more column in sparse output to calculate label count. sparse_ada_grad_schema in tensornet-tools is not update-to-date. Or you don't need to enable target_columns.

normally, enable use_cvm_ is helpful to improve the model efficiency?
